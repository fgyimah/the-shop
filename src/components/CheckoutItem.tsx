import React from 'react';
import styled from 'styled-components';
import { useDispatch } from 'react-redux';
import { Product } from '../@types';
import { addItemToCart, clearItemFromCart, removeItemFromCart } from '../store/cart.reducer';

interface Props {
    cartItem: { product: Product; quantity: number };
}

const CheckoutItem: React.FC<Props> = ({ cartItem }) => {
    const { product, quantity } = cartItem;
    const dispatch = useDispatch();

    const increaseCount = () => {
        dispatch(addItemToCart(product));
    };

    const decreaseCount = () => {
        dispatch(removeItemFromCart(product));
    };

    const removeItem = () => {
        dispatch(clearItemFromCart(product));
    };

    return (
        <StyledDiv>
            <div className="checkout-item">
                <div className="image-container">
                    <img src={product.imageUrl} alt="item" />
                </div>
                <span className="name">{product.name}</span>
                <span className="quantity">
                    <div className="arrow" onClick={decreaseCount}>
                        &#10094;
                    </div>
                    <span className="value">{quantity}</span>
                    <div className="arrow" onClick={increaseCount}>
                        &#10095;
                    </div>
                </span>
                <span className="price">{product.price}</span>
                <div className="remove-button" onClick={removeItem}>
                    &#10005;
                </div>
            </div>
        </StyledDiv>
    );
};

const StyledDiv = styled.div`
    .checkout-item {
        width: 100%;
        display: flex;
        min-height: 100px;
        border-bottom: 1px solid darkgrey;
        padding: 15px 0;
        font-size: 20px;
        align-items: center;

        .image-container {
            width: 23%;
            padding-right: 15px;

            img {
                width: 100%;
                height: 100%;
            }
        }
        .name,
        .quantity,
        .price {
            width: 23%;
        }

        .quantity {
            display: flex;

            .arrow {
                cursor: pointer;
            }

            .value {
                margin: 0 10px;
            }
        }

        .remove-button {
            padding-left: 12px;
            cursor: pointer;
        }
    }
    @media screen and (max-width: 720px) {
        .checkout-item {
            flex-direction: column;
            align-items: center;
            text-align: center;
        }
    }
`;

export default CheckoutItem;
