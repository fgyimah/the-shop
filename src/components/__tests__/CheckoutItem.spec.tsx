import { render } from '@testing-library/react';
import * as reactRedux from 'react-redux';
import CheckoutItem from '../CheckoutItem';

describe('Component Test - CheckoutItem', () => {
    const useSelectorMock = jest.spyOn(reactRedux, 'useSelector');
    const useDispatchMock = jest.spyOn(reactRedux, 'useDispatch');

    beforeEach(() => {
        useSelectorMock.mockClear();
        useDispatchMock.mockClear();
    });

    const data = {
        product: {
            id: '1',
            name: 'fake-product',
            description: 'fake description',
            price: 99,
            imageUrl: '',
        },
        quantity: 5,
    };

    it('matches snapshot', () => {
        const { asFragment } = render(<CheckoutItem cartItem={data} />);
        expect(asFragment()).toMatchSnapshot();
    });

    it('renders the quantity inside the checkout item', () => {
        const { getByTestId } = render(<CheckoutItem cartItem={data} />);
        expect(getByTestId('product-quantity').innerHTML).toMatch('5');
    });
});
