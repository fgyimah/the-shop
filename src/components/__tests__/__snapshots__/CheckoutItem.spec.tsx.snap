// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component Test - CheckoutItem matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-bdfBwQ hrReyS"
  >
    <div
      class="checkout-item"
    >
      <div
        class="image-container"
      >
        <img
          alt="item"
          src=""
        />
      </div>
      <span
        class="name"
      >
        fake-product
      </span>
      <span
        class="quantity"
      >
        <div
          class="arrow"
          data-testid="btn-decrease-count"
        >
          ❮
        </div>
        <span
          class="value"
          data-testid="product-quantity"
        >
          5
        </span>
        <div
          class="arrow"
          data-testid="btn-increase-count"
        >
          ❯
        </div>
      </span>
      <span
        class="price"
        data-testid="product-price"
      >
        99
      </span>
      <div
        class="remove-button"
        data-testid="btn-remove"
      >
        ✕
      </div>
    </div>
  </div>
</DocumentFragment>
`;
